### PR TITLE
chore(rln-relay): add isReady check

### PR DESCRIPTION
**Pull request by rymnc**
# Description

Adds an `isReady` proc to waku_node, waku_rln_relay  

# Changes

- [x] WakuNode.isReady => if all protocols are ready, returns true
- [x] WakuRlnRelay.isReady => if group manager is ready, returns true

## Issue

Addresses 25. #1906

### CODE REVIEW

**Comment by rymnc** [comment]
cc: @alrevuelta i think this pr should be separate from the one that creates an api endpoint to reduce cognitive load - wdyt?

**Review by alrevuelta — COMMENTED** [review]
@rymnc agree, we can have 2 PRs.

Not sure having a `isReady` variable is a good idea. I would have just a function `isReady()` that check the difference between the last processed block in the rln tree and the latest head block in the blockchain (perhaps allowing 1 or 2 blocks of difference).

As it is, you may set `g.isReady = true` but at any point in time you can lose sync and unless im missing something you wont set isReady back to false. Imagine for example that the eth node falls behind sync.

tldr: `isReady()` function but no stored `isReady` state. This state is calculated upon request. wdyt?

**Comment by rymnc** [comment]
> @rymnc agree, we can have 2 PRs.
> 
> Not sure having a `isReady` variable is a good idea. I would have just a function `isReady()` that check the difference between the last processed block in the rln tree and the latest head block in the blockchain (perhaps allowing 1 or 2 blocks of difference).
> 
> As it is, you may set `g.isReady = true` but at any point in time you can lose sync and unless im missing something you wont set isReady back to false. Imagine for example that the eth node falls behind sync.
> 
> tldr: `isReady()` function but no stored `isReady` state. This state is calculated upon request. wdyt?

I think that's overkill - when we are out of sync, we do set isReady to false (check the ethRpc.onDisconnect)

**Comment by alrevuelta** [comment]
> I think that's overkill - when we are out of sync, we do set isReady to false (check the ethRpc.onDisconnect)

yep but there are other ways you can get out of sync without ethRpc.onDisconnect. For example, if the eth node gets out of sync, then you will fall behind while staying connected. You assume that once you are in sync for the first time, you stay in sync forever (unless disconnected). What happens also if your node can't process all the membership insertions realtime and falls behind?

**Comment by rymnc** [comment]
> > I think that's overkill - when we are out of sync, we do set isReady to false (check the ethRpc.onDisconnect)
> 
> yep but there are other ways you can get out of sync without ethRpc.onDisconnect. For example, if the eth node gets out of sync, then you will fall behind while staying connected. You assume that once you are in sync for the first time, you stay in sync forever (unless disconnected). What happens also if your node can't process all the membership insertions realtime and falls behind?

hmm, how would it be possible to detect if the eth node itself is out of sync without using multiple providers?

>  What happens also if your node can't process all the membership insertions realtime and falls behind?
should we set it to false then during inserts? and set it back to true right after?

**Comment by alrevuelta** [comment]
> hmm, how would it be possible to detect if the eth node itself is out of sync without using multiple providers?

afaik even if the eth node is out of sync, it knows that its out sync. meaning that it knows the latest head block known by the blockchain and the latest known block by the node. I think this is the rpc [link](https://docs.infura.io/networks/ethereum/json-rpc-methods/eth_syncing#returns)

I guess you will also notice indirectly. If the eth node falls behind sync, it wont give you the latest blocks, so you will fall behind sync, seeing that the latestHead is way beyond your latest localRlnLatestProcessedBlock.

**Comment by alrevuelta** [comment]
> if the eth node falls behind sync, would it still emit the newHead events? unsure about this behaviour - i do agree about calculating every time isReady is called though

have never tried empirically. I would bet that it stops emitting events, and you may fall behind sync without noticing.

**Comment by rymnc** [comment]
> > if the eth node falls behind sync, would it still emit the newHead events? unsure about this behaviour - i do agree about calculating every time isReady is called though
> 
> have never tried empirically. I would bet that it stops emitting events, and you will fall behind sync without noticing.

yes so then there are 2 ways we will be out of sync -
1. eth node out of sync
2. we do not process insertions in time 

we cannot account for 1 unless we use multiple providers
for 2, we can do this by having the isReady proc compute the difference between latestBlockSeen and latestBlockProcessed, as you said. wdyt?

**Comment by alrevuelta** [comment]
> we cannot account for 1 unless we use multiple providers

why? an eth node out of sync will know its out of sync. so with just one node you should know that its out of sync. or?

afaik we can address both 1 and 2 without multiple providers unless im missing something.

**Comment by rymnc** [comment]
> > we cannot account for 1 unless we use multiple providers
> 
> why? an eth node out of sync will know its out of sync. so with just one node you should know that its out of sync. or?
> 
> afaik we can address both 1 and 2 without multiple providers unless im missing something.

maybe are you suggesting that we make the rpc call every time the `isReady` proc is called?

**Review by Ivansete-status — APPROVED** [review]
LGTM! Super interesting idea!

In further PRs, I think we will need to add a `/healthcheck` endpoint as a valid REST path ( cc @NagyZoltanPeter )

**Comment by alrevuelta** [comment]
> maybe are you suggesting that we make the rpc call every time the isReady proc is called?

Yes exactly, perhaps this: https://docs.infura.io/networks/ethereum/json-rpc-methods/eth_syncing#returns

I'm fine with leaving a TODO for this (your point 1). But would like to address 2. here.

> In further PRs, I think we will need to add a /healthcheck endpoint as a valid REST path ( cc @NagyZoltanPeter )

tracked here:
https://github.com/waku-org/nwaku/issues/1988

**Comment by NagyZoltanPeter** [comment]
I like the idea of having /healthcheck!

**Review by SionoiS — APPROVED** [review]
LGTM

**Comment by rymnc** [comment]
@alrevuelta I hope cbb266ae covers the requirement (1 & 2)

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim [hunk 1]
```diff
@@ -528,3 +535,27 @@ method stop*(g: OnchainGroupManager): Future[void] {.async.} =
...
+  let ethRpc = g.ethRpc.get()
+
+  try:
+    let syncing = await ethRpc.provider.eth_syncing()
+    return syncing.getBool()
+  except CatchableError:
+    error "failed to get the syncing status", error = getCurrentExceptionMsg()
+    return false
+
+method isReady*(g: OnchainGroupManager): Future[bool] {.async,gcsafe.} =
```
Comment (alrevuelta) [inline waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:549]: perhaps its cleaner to run the checks for `true` all together?

```
if g.ethRpc.isSome() && g.lastSeenBlockHead !=0 && g.latestProcessedBlock >= g.lastSeenBlockHead && !await g.isSyncing():
  return true
return false
```

maybe just a personal prefernce, just a sugerence.

**Review by alrevuelta — COMMENTED** [review]
thanks! left a comment of a possible edge case, let me know if im missing something.

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim [hunk 2]
```diff
@@ -73,6 +73,8 @@ type
     # in event of a reorg. we store 5 in the buffer. Maybe need to revisit this,
     # because the average reorg depth is 1 to 2 blocks.
     validRootBuffer*: Deque[MerkleNode]
+    # this variable tracks the last seen head
+    lastSeenBlockHead*: BlockNumber
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:77]: right, I'll just move it to the newHeadCallback, thanks

[the remark this answers is not in the document]
Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:77]: Addressed in 3c9d4bef

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim (hunk 1 above)
Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:548]: Yeah, I did think of this, imo the way it is right now is slightly more readable

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (alrevuelta) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:77]: > So the lastSeenBlockHead should be the last head block known by the blockchain.

I think its safer to get this from `eth_blockNumber` endpoint rather than relying on the newHeadcallback. If the eth node fails to keep you updated with the latest head (but without disconnecting) then your lastSeenBlockHead will be outdated.

But if you use `eth_blockNumber`, since its req/rep, you ensure that your latest head is correct. If then node fails to reply, then its not ready.

[the remark this answers is not in the document]
Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim]: Addressed in 15bfde44

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim [hunk 3]
```diff
@@ -551,10 +553,12 @@ method isReady*(g: OnchainGroupManager): Future[bool] {.async,gcsafe.} =
   if g.ethRpc.isNone():
     return false
 
-  if g.lastSeenBlockHead == 0:
-    return false
+  let currentBlock = cast[BlockNumber](await g.ethRpc
```
Comment (alrevuelta) [inline waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:556]: guess if this fails an exceptions is thrown? should we catch it and return `false` if so. guess if we dont catch it the node will crash?

**Review by alrevuelta — APPROVED** [review]
lgtm!
